src/mito/src/engine/procedure/drop.rs

```diff
@@ -0,0 +1,227 @@
+            DropTableState::Prepare => self.on_prepare(),
+            DropTableState::CloseRegions => self.on_close_regions().await,
+        }
+    }
+
+    fn dump(&self) -> Result<String> {
+        let json = serde_json::to_string(&self.data).context(ToJsonSnafu)?;
+        Ok(json)
+    }
+
+    fn lock_key(&self) -> LockKey {
+        let table_ref = self.data.table_ref();
+        let info = self.table.table_info();
+        let keys = info
+            .meta
+            .region_numbers
+            .iter()
+            .map(|number| format!("{table_ref}/region-{number}"));
+        LockKey::new(keys)
+    }
+}
+
+impl<S: StorageEngine> DropMitoTable<S> {
+    const TYPE_NAME: &str = "mito::DropMitoTable";
+
+    /// Returns a new [DropMitoTable].
+    pub(crate) fn new(
+        request: DropTableRequest,
+        engine_inner: Arc<MitoEngineInner<S>>,
+    ) -> Result<Self> {
+        let data = DropTableData {
+            state: DropTableState::Prepare,
+            request,
+        };
+        let table_ref = data.table_ref();
+        let table =
+            engine_inner
+                .get_mito_table(&table_ref)
+                .with_context(|| TableNotFoundSnafu {
+                    table_name: table_ref.to_string(),
+                })?;
+
+        Ok(DropMitoTable {
+            data,
+            engine_inner,
+            table,
+        })
+    }
+
+    /// Register the loader of this procedure to the `procedure_manager`.
+    ///
+    /// # Panics
+    /// Panics on error.
+    pub(crate) fn register_loader(
+        engine_inner: Arc<MitoEngineInner<S>>,
+        procedure_manager: &dyn ProcedureManager,
+    ) {
+        procedure_manager
+            .register_loader(
+                Self::TYPE_NAME,
+                Box::new(move |data| {
+                    Self::from_json(data, engine_inner.clone()).map(|p| Box::new(p) as _)
+                }),
+            )
+            .unwrap()
+    }
+
+    /// Recover the procedure from json.
+    fn from_json(json: &str, engine_inner: Arc<MitoEngineInner<S>>) -> Result<Self> {
+        let data: DropTableData = serde_json::from_str(json).context(FromJsonSnafu)?;
+        let table_ref = data.table_ref();
+        let table =
+            engine_inner
+                .get_mito_table(&table_ref)
+                .with_context(|| TableNotFoundSnafu {
+                    table_name: table_ref.to_string(),
+                })?;
+
+        Ok(DropMitoTable {
+            data,
+            engine_inner,
+            table,
+        })
+    }
+
+    /// Prepare table info.
+    fn on_prepare(&mut self) -> Result<Status> {
+        self.data.state = DropTableState::CloseRegions;
+
+        Ok(Status::executing(true))
+    }
+
+    /// Close all regions.
+    async fn on_close_regions(&mut self) -> Result<Status> {
+        // Close the table to close all regions. Closing a region idempotent.
+        self.table.close().await.map_err(Error::from_error_ext)?;
```

Comment:
Consider removing from engine_inner.tables first before closing?  Otherwise, users may access a table that is half closed.